### PR TITLE
5.0 C test of task with affinity clause

### PR DESCRIPTION
**Pull request by jhdavis8**
These are tests of task affinity from 5.0 in C. Included is a device and host version. Clang will compile the host test, and it passes, but it crashes with the device test:

```
nvlink error   : Undefined reference to '__kmpc_omp_reg_task_with_affinity' in '/tmp/test_task_affinity_device-2e2189.cubin'
clang-13: error: nvlink command failed with exit code 255 (use -v to see invocation)
clang version 13.0.0 (https://github.com/llvm/llvm-project.git 3bc899b4de74d395b03d3969d700aac71f89bc12)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/ums/stf010/llvm/13.0.0-20210615/bin
clang-13: note: diagnostic msg: Error generating preprocessed source(s).
```

The other Summit C compilers don't support the affinity clause yet.

### CODE REVIEW

File: tests/5.0/task/test_task_affinity_device.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  A = (int*) omp_target_alloc(sizeof(int)*N, t);
+
+#pragma omp target is_device_ptr(A) device(t) map(from: B[0:N])
+  {
+#pragma omp task depend(out: B) shared(B) affinity(A[0:N])
+    {
+      for (int i = 0; i < N; i++) {
+        A[i] = 0;
+      }
+      B = (int*) malloc(sizeof(int)*N);
```
Comment (jhdavis8) [inline tests/5.0/task/test_task_affinity_device.c:39]: Try just allocating B the same way A is allocated.

Comment (jhdavis8) [reply tests/5.0/task/test_task_affinity_device.c:39]: Update on this, switching the way B is allocated did not affect the results for Clang. Still fails to compile with the same error message.

File: tests/5.0/task/test_task_affinity_device.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+int test_task_affinity() {
+  OMPVV_INFOMSG("test_task_affinity");
+  int errors = 0;
+  int* A;
+  int* B;
+  int t = omp_get_default_device();
+
+  A = (int*) omp_target_alloc(sizeof(int)*N, t);
+
+#pragma omp target is_device_ptr(A) device(t) map(from: B[0:N])
```
Comment (jhdavis8) [inline tests/5.0/task/test_task_affinity_device.c:32]: Try allocing B instead of mapping

Comment (jhdavis8) [reply tests/5.0/task/test_task_affinity_device.c:32]: Try moving the target inside the task

Comment (jhdavis8) [reply tests/5.0/task/test_task_affinity_device.c:32]: Compiler error output when changed so the tasks generate target regions:
```
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13 -cc1 -triple powerpc64le-unknown-linux-gnu -emit-llvm-bc -emit-llvm-uselists -disable-free -disable-llvm-verifier
 -discard-value-names -main-file-name test_task_affinity_device.c -mrelocation-model static -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu ppc64le -mfloat-ab
i hard -target-abi elfv2 -debugger-tuning=gdb -fcoverage-compilation-dir=/ccs/home/jhdavis/sollve-project/sollve_vv -resource-dir /autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/lib/clang/13.0.0 -I ./o
mpvv -D __NO_MATH_INLINES -U __SSE2_MATH__ -U __SSE_MATH__ -D VERBOSE_MODE=1 -I/sw/summit/cuda/10.1.243/include -c-isystem /sw/summit/ums/stf010/llvm/13.0.0-20210713/include -cxx-isystem /sw/summit/ums/stf010/llvm
/13.0.0-20210713/include -internal-isystem /autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/lib/clang/13.0.0/include/ppc_wrappers -internal-isystem /autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-202
10713/lib/clang/13.0.0/include -internal-isystem /usr/local/include -internal-isystem /sw/summit/gcc/9.1.0-alpha+20190716/lib/gcc/powerpc64le-unknown-linux-gnu/9.1.0/../../../../powerpc64le-unknown-linux-gnu/inclu
de -internal-externc-isystem /include -internal-externc-isystem /usr/include -internal-isystem /autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/lib/clang/13.0.0/include/ppc_wrappers -internal-isystem /a
utofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/lib/clang/13.0.0/include -internal-isystem /usr/local/include -internal-isystem /sw/summit/gcc/9.1.0-alpha+20190716/lib/gcc/powerpc64le-unknown-linux-gnu/9
.1.0/../../../../powerpc64le-unknown-linux-gnu/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -O3 -fdebug-compilation-dir=/ccs/home/jhdavis/sollve-project/sollve_vv -ferror-limit
 19 -fopenmp -fno-signed-char -fgnuc-version=4.2.1 -vectorize-loops -vectorize-slp -disable-llvm-passes -fopenmp-targets=nvptx64-nvidia-cuda -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/test_task_affinity_devic
e-96d5dd.bc -x c /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/task/test_task_affinity_device.c
1.      <eof> parser at end of file
2.      Per-file LLVM IR generation
3.      /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/task/test_task_affinity_device.c:23:5: Generating code for declaration 'test_task_affinity'
4.      /autofs/nccs-svm1_home1/jhdavis/sollve-project/sollve_vv/tests/5.0/task/test_task_affinity_device.c:33:5: LLVM IR generation of compound statement ('{}')
 #0 0x0000000011773014 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11773014)
 #1 0x0000000011773414 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #2 0x00000000117703d8 llvm::sys::RunSignalHandlers() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x117703d8)
 #3 0x000000001177385c SignalHandler(int) Signals.cpp:0:0
 #4 0x00002000000504d8  0x4d8 clang::CodeGen::CodeGenFunction::EmitDeclRefLValue(clang::DeclRefExpr const*)
 #5 0x00002000000504d8 
 #6 0x00002000000504d8 clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*) (+0x4d8)
 #7 0x0000000011cc0e70 clang::CodeGen::CodeGenFunction::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-
13+0x11cc0e70)
 #8 0x0000000011cb55b4 (anonymous namespace)::ScalarExprEmitter::VisitDeclRefExpr(clang::DeclRefExpr*) CGExprScalar.cpp:0:0
 #9 0x0000000011cbd428 (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) CGExprScalar.cpp:0:0
#10 0x0000000011cfac94 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11cfac94)
#11 0x0000000011cee990 clang::CodeGen::CodeGenFunction::EmitScalarInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/cl
ang-13+0x11cee990)
#12 0x0000000011cee900 clang::CodeGen::CodeGenFunction::EmitExprAsInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/cl
ang-13+0x11cee900)
#13 0x0000000011c9a104 clang::CodeGen::CodeGenFunction::EmitAutoVarInit(clang::CodeGen::CodeGenFunction::AutoVarEmission const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11c9a104
)
#14 0x0000000011c9ef64 clang::CodeGen::CodeGenFunction::EmitVarDecl(clang::VarDecl const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11c9ef64)
#15 0x0000000011c9c5f4 (anonymous namespace)::OMPLexicalScope::OMPLexicalScope(clang::CodeGen::CodeGenFunction&, clang::OMPExecutableDirective const&, llvm::Optional<llvm::omp::Directive>, bool) CGStmtOpenMP.cpp:0
:0
#16 0x0000000011c981ac emitCommonOMPTargetDirective(clang::CodeGen::CodeGenFunction&, clang::OMPExecutableDirective const&, clang::CodeGen::RegionCodeGenTy const&) CGStmtOpenMP.cpp:0:0
#17 0x0000000011a6d4ec clang::CodeGen::CodeGenFunction::EmitOMPTargetDirective(clang::OMPTargetDirective const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a6d4ec)
#18 0x0000000011a6fa60 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a6fa60)
#19 0x0000000011a7e22c clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a7e22c)
#20 0x0000000011a4c148 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a4c148)
#21 0x0000000011a4c544 void clang::CodeGen::RegionCodeGenTy::CallbackFn<clang::CodeGen::CodeGenFunction::EmitOMPTaskDirective(clang::OMPTaskDirective const&)::$_80>(long, clang::CodeGen::CodeGenFunction&, clang::C
odeGen::PrePostActionTy&) CGStmtOpenMP.cpp:0:0
#22 0x0000000011a4bb34 clang::CodeGen::RegionCodeGenTy::operator()(clang::CodeGen::CodeGenFunction&) const (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a4bb34)
#23 0x0000000011a97e6c void clang::CodeGen::RegionCodeGenTy::CallbackFn<clang::CodeGen::CodeGenFunction::EmitOMPTaskBasedDirective(clang::OMPExecutableDirective const&, llvm::omp::Directive, clang::CodeGe[51/1695]
CodeGenTy const&, llvm::function_ref<void (clang::CodeGen::CodeGenFunction&, llvm::Function*, clang::CodeGen::OMPTaskDataTy const&)> const&, clang::CodeGen::OMPTaskDataTy&)::$_74>(long, clang::CodeGen::CodeGenFunc
tion&, clang::CodeGen::PrePostActionTy&) CGStmtOpenMP.cpp:0:0
#24 0x0000000011d7d754 (anonymous namespace)::CGOpenMPRegionInfo::EmitBody(clang::CodeGen::CodeGenFunction&, clang::Stmt const*) CGOpenMPRuntime.cpp:0:0
#25 0x0000000011a966c4 clang::CodeGen::CodeGenFunction::GenerateCapturedStmtFunction(clang::CapturedStmt const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a966c4)
#26 0x0000000011db8460 clang::CodeGen::CGOpenMPRuntime::emitTaskOutlinedFunction(clang::OMPExecutableDirective const&, clang::VarDecl const*, clang::VarDecl const*, clang::VarDecl const*, llvm::omp::Directive, cla
ng::CodeGen::RegionCodeGenTy const&, bool, unsigned int&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11db8460)
#27 0x0000000011a5adfc clang::CodeGen::CodeGenFunction::EmitOMPTaskBasedDirective(clang::OMPExecutableDirective const&, llvm::omp::Directive, clang::CodeGen::RegionCodeGenTy const&, llvm::function_ref<void (clang:
:CodeGen::CodeGenFunction&, llvm::Function*, clang::CodeGen::OMPTaskDataTy const&)> const&, clang::CodeGen::OMPTaskDataTy&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a5adfc)
#28 0x0000000011d83660 clang::CodeGen::CodeGenFunction::EmitOMPTaskDirective(clang::OMPTaskDirective const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11d83660)
#29 0x0000000011a779f8 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a779f8)
#30 0x0000000011a7933c clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin
/clang-13+0x11a7933c)
#31 0x0000000011a4c1d4 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clan
g-13+0x11a4c1d4)
#32 0x0000000011a56524 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11a56524)
#33 0x0000000011aafbbc clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11aafbbc)
#34 0x0000000011ad07a8 clang::CodeGen::CodeGenModule::EmitDeferred() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11ad07a8)
#35 0x0000000011ac9058 clang::CodeGen::CodeGenModule::Release() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11ac9058)
#36 0x0000000011abb604 (anonymous namespace)::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) ModuleBuilder.cpp:0:0
#37 0x0000000011ab9cb0 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11ab9cb0)
#38 0x0000000012629304 clang::ParseAST(clang::Sema&, bool, bool) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x12629304)
#39 0x0000000012625340 clang::ASTFrontendAction::ExecuteAction() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x12625340)
#40 0x00000000131d9bb0 clang::CodeGenAction::ExecuteAction() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x131d9bb0)
#41 0x0000000011ffb568 clang::FrontendAction::Execute() (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11ffb568)
#42 0x0000000012623724 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x12623724)
#43 0x0000000011ffae14 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11ffae14)
#44 0x0000000011f72988 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x11f72988)
#45 0x00000000120c6578 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#46 0x000000001060c174 main (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x1060c174)
#47 0x000000001060a150 generic_start_main.isra.0 libc-start.c:0:0
#48 0x0000000010609fd8 __libc_start_main (/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13+0x10609fd8)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x44)[0x11773014]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11773414]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN4llvm3sys17RunSignalHandlersEv+0xc8)[0x117703d8]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x1177385c]
[0x2000000504d8]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction17EmitDeclRefLValueEPKNS_11DeclRefExprE+0x1890)[0x11cc0e70]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction10EmitLValueEPKNS_4ExprE+0x294)[0x11cb55b4]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction17EmitCheckedLValueEPKNS_4ExprENS1_13TypeCheckKindE+0x88)[0x11cbd428]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11cfac94]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11cee990]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction14EmitScalarExprEPKNS_4ExprEb+0x60)[0x11cee900]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction14EmitScalarInitEPKNS_4ExprEPKNS_9ValueDeclENS0_6LValueEb+0x204)[0x11c9a104]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction14EmitExprAsInitEPKNS_4ExprEPKNS_9ValueDeclENS0_6LValueEb+0x374)[0x11c9ef64]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction15EmitAutoVarInitERKNS1_15AutoVarEmissionE+0x974)[0x11c9c5f4]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction11EmitVarDeclERKNS_7VarDeclE+0x12c)[0x11c981ac]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11a6d4ec]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11a6fa60]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction22EmitOMPTargetDirectiveERKNS_18OMPTargetDirectiveE+0x3c)[0x11a7e22c]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction8EmitStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0x6f8)[0x11a4c148]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction14EmitSimpleStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0x224)[0x11a4c544]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction8EmitStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0xe4)[0x11a4bb34]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11a97e6c]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZNK5clang7CodeGen15RegionCodeGenTyclERNS0_15CodeGenFunctionE+0xd4)[0x11d7d754]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11a966c4]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x11db8460]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction28GenerateCapturedStmtFunctionERKNS_12CapturedStmtE+0x7fc)[0x11a5adfc]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CGOpenMPRuntime24emitTaskOutlinedFunctionERKNS_22OMPExecutableDirectiveEPKNS_7VarDeclES7_S7_N4llvm3omp9DirectiveERKNS0_15
RegionCodeGenTyEbRj+0x240)[0x11d83660]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction25EmitOMPTaskBasedDirectiveERKNS_22OMPExecutableDirectiveEN4llvm3omp9DirectiveERKNS0_15RegionCodeGenTyERKN
S5_12function_refIFvRS1_PNS5_8FunctionERKNS0_13OMPTaskDataTyEEEERSF_+0x1758)[0x11a779f8]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction20EmitOMPTaskDirectiveERKNS_16OMPTaskDirectiveE+0x3dc)[0x11a7933c]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction8EmitStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0x784)[0x11a4c1d4]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction28EmitCompoundStmtWithoutScopeERKNS_12CompoundStmtEbNS0_12AggValueSlotE+0x2a4)[0x11a56524]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen15CodeGenFunction12GenerateCodeENS_10GlobalDeclEPN4llvm8FunctionERKNS0_14CGFunctionInfoE+0x4dc)[0x11aafbbc]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen13CodeGenModule28EmitGlobalFunctionDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x288)[0x11ad07a8]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen13CodeGenModule20EmitGlobalDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x1e8)[0x11ac9058]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen13CodeGenModule12EmitDeferredEv+0x384)[0x11abb604]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang7CodeGen13CodeGenModule7ReleaseEv+0x70)[0x11ab9cb0]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x12629304]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x12625340]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang8ParseASTERNS_4SemaEbb+0x320)[0x131d9bb0]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang17ASTFrontendAction13ExecuteActionEv+0xd8)[0x11ffb568]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang13CodeGenAction13ExecuteActionEv+0x114)[0x12623724]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang14FrontendAction7ExecuteEv+0xa4)[0x11ffae14]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x438)[0x11f72988]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x6d8)[0x120c6578]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x704)[0x1060c174]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13[0x1060a150]
/autofs/nccs-svm1_sw/summit/ums/stf010/llvm/13.0.0-20210713/bin/clang-13(main+0x3118)[0x10609fd8]
/usr/lib64/libc.so.6(+0x25200)[0x2000004a5200]
/usr/lib64/libc.so.6(__libc_start_main+0xc4)[0x2000004a53f4]
clang-13: error: unable to execute command: Segmentation fault (core dumped)
clang-13: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 13.0.0 (https://github.com/llvm/llvm-project.git 46e89708170c40e8cf0305b6de048ca879f43aab)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/ums/stf010/llvm/13.0.0-20210713/bin
clang-13: note: diagnostic msg: Error generating preprocessed source(s).
```

**Comment by spophale** [comment]
@jhdavis8 Does the new device test work with LLVM trunk ?

**Comment by jhdavis8** [comment]
In short, no. The crash output I posted above also appears using the newer 7/14 build of Clang on Summit. I was using the previous build at the time.

File: tests/5.0/task/test_task_affinity_device.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    OMPVV_TEST_AND_SET_VERBOSE(errors, B[i] != i*2);
+    OMPVV_TEST_AND_SET_VERBOSE(errors, A[i] != 0);
+  }
+
+  return errors;
+}
+
+int main() {
+  int errors = 0;
+
```
Comment (jhdavis8) [inline tests/5.0/task/test_task_affinity_device.c:68]: Add OMPVV_TEST_OFFLOADING here

**Review by tmh97 — APPROVED** [review]

**Comment by spophale** [comment]
I want to make sure the bug was reported before we close this PR.

**Review by spophale — APPROVED** [review]

File: tests/5.0/task/test_task_affinity_device.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  int errors = 0;
+  int* A;
+  int* B;
+  int t = omp_get_default_device();
+
+  A = (int*) omp_target_alloc(sizeof(int)*N, t);
+
+#pragma omp task depend(out: B) shared(B) affinity(A[0:N])
+    {
+#pragma omp target is_device_ptr(A) device(t) map(from: B[0:N])
```
Comment (tmh97) [inline tests/5.0/task/test_task_affinity_device.c:34]: alloc b outside of target region and see if there is still a crash

**Comment by tob2** [comment]
Testings with **GCC 12**:

* Both files compile
* test_task_affinity.c –  runs successfully.
* test_task_affinity_device.c fails at runtime with (-fsanitize=address,undefined; only tried with host fallback):
```
test_task_affinity_device.c:51:16: runtime error: store to null pointer of type 'int'

==811783==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000402954 bp 0x7fff735f2ae0 sp 0x7fff735f2a50 T0)
==811783==The signal is caused by a WRITE memory access.
==811783==Hint: address points to the zero page.
    #0 0x402954 in test_task_affinity_device.c:51
```

I am not sure I understand the code – 'b' is not allocated on the host & might or might not be NULL initialized as stack variable. Thus, it is completely unclear how that should work.

**Comment by spophale** [comment]
@tmh97 @nolanbaker31 : re-review. 
@tob2 : Does this work with gcc 12 (does not with 11 on Summit).

Now the device test compiles and works with LLVM on Summit.

**Comment by spophale** [comment]
LLVM error:
`clang-14: warning: unknown CUDA version: version.txt: 11.0.228.; assuming the latest supported version 10.1 [-Wunknown-cuda-version]
nvlink error   : Undefined reference to '__kmpc_omp_reg_task_with_affinity' in '/tmp/test_task_affinity_device_tasksondevice-cb08e4.cubin'
clang-14: error: nvlink command failed with exit code 255 (use -v to see invocation)
clang version 14.0.0 (https://github.com/llvm/llvm-project.git 7bdce6bcbda3a8b7dcdac6a7d8fb1083912830d7)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/ums/stf010/llvm/13.0.0-20210810/bin`

File: tests/5.0/task/test_task_affinity.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    B = (int*) malloc(sizeof(int)*N);
+    for (int i = 0; i < N; i++) {
+      B[i] = A[i];
+    }
+  }
+
+#pragma omp task depend(in: B) shared(B) affinity(A[0:N])
+  {
+    for (int i = 0; i < N; i++) {
+      B[i] = i*2;
```
Comment (spophale) [inline tests/5.0/task/test_task_affinity.c:44]: use A

**Comment by spophale** [comment]
@tob2 can you re-try test_task_affinity_device.c with GCC 12 ?

**Comment by tob2** [comment]
> @tob2 can you re-try test_task_affinity_device.c with GCC 12 ?

Works now :-)

I did get memory leak warnings with ASAN (-fsanitize=address), hence, I created PR #415 (now merged).

**Pull request merged**